spring-cloud-gateway-core/src/main/java/org/springframework/cloud/gateway/filter/factory/rewrite/GzipMessageBodyResolver.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright 2013-2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.gateway.filter.factory.rewrite;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.util.zip.GZIPInputStream;
+import java.util.zip.GZIPOutputStream;
+
+import org.springframework.core.io.buffer.DataBuffer;
+import org.springframework.util.FileCopyUtils;
+
+public class GzipMessageBodyResolver implements MessageBodyDecoder, MessageBodyEncoder {
+
+	@Override
+	public String encodingType() {
+		return "gzip";
+	}
+
+	@Override
+	public byte[] decode(byte[] encoded) {
+		try {
+			ByteArrayInputStream bis = new ByteArrayInputStream(encoded);
+			GZIPInputStream gis = new GZIPInputStream(bis);
```

Comment:
I'm afraid anything using input or output stream is by definition blocking and as such can not be used in gateway

Reply:
This call shouldn't block, cause underlying stream in ByteArray based and all bytes are already available at this point.  I believe that call would've been blocking in case if I've passed response inputStream instead of byte array and tried to read chunked response for example, however it's not the case here. In order to make any response body modification by definition we need to read the whole response first, it doesn't matter if it's encoded or not, cause modiciation function anyway requires full response. Same applies to encode method. Thus in this particular case it's just streaming data from one byte array to another with encoding/decoding it. GzipInputStream impementation is based on [Inflater](https://docs.oracle.com/javase/7/docs/api/java/util/zip/Inflater.html) which blocks only if there is not enough input at the moment to decode data. GzipOutputStream is based on [Deflater](https://docs.oracle.com/javase/7/docs/api/java/util/zip/Deflater.html). 

I understand that for big responses it can take some time to allocate enough memory and encode/decode data and probably running this on the same scheduler might be not the best idea, I can try making it run on a separate scheduler, but wouldn't same problem then apply to the prior call where you are reading the whole response and casting it to inClass? 

If you still think that InputSteam/OutputStream is not applicable in that scenario, I can try substituting it with something different, like [ZlibDecoder](https://netty.io/4.0/api/io/netty/handler/codec/compression/ZlibDecoder.html) from netty (not sure though if implementation should depend on netty). I wasn't able to find any async implementation in Spring WebFlux.

Reply:
We already have a hard dependency on netty.

Reply:
After some digging I can see that netty `ZlibDecoder` is also internally implemented via same jdk `Inflater` as GZIPInputSteam and basically does the same thing. So not sure what's the benefit of using it instead of standard jdk one.
Can you please clarify what do you mean by blocking? As mentioned earlier, I don't think there are any blocking operations in that piece of code as there are no IO(since these streams are byte array based) or concurrent access.
I can try moving this call to a separate scheduler if that's your concern.

Reply:
@simonbasle or @violetagg thoughts?

Reply:
@simonbasle, @violetagg or @spencergibb can you please review?

Reply:
if we're guaranteed that all the bytes are in memory, it is indeed a bit safer. I would advise to offset that compression work on a separate `Scheduler` to be on the safe side with large payloads. Compression is CPU bound, so `parallel()` sounds like a good choice.

Reply:
@korektur can we do as Simon and you have mentioned and move to the `parallel` scheduler?

Reply:
@spencergibb done